server/src/main/java/org/elasticsearch/script/field/WriteField.java

```diff
@@ -670,24 +857,34 @@ protected void resolveDepthFlat() {
      */
     @SuppressWarnings("unchecked")
     protected void createDepth() {
-        container = rootSupplier.get();
-
+        Map<String, Object> mapContainer = rootSupplier.get();
         String[] segments = path.split("\\.");
         for (int i = 0; i < segments.length - 1; i++) {
             String segment = segments[i];
-            Object value = container.getOrDefault(segment, MISSING);
+            Object value = mapContainer.getOrDefault(segment, MISSING);
             if (value instanceof Map<?, ?> map) {
-                container = (Map<String, Object>) map;
+                mapContainer = (Map<String, Object>) map;
             } else if (value == MISSING) {
                 Map<String, Object> next = Maps.newHashMapWithExpectedSize(4);
-                container.put(segment, next);
-                container = next;
+                mapContainer.put(segment, next);
```

Comment:
This logic does not seem to handle indices in the path name. If trying to set the value at `a.b.c.0` when that field does not exist, the result will be a map near the leaf of the tree that contains the key `"0"` instead of an array.

Reply:
> If trying to set the value at a.b.c.0 when that field does not exist, the result will be a map near the leaf of the tree that contains the key "0" instead of an array.

This is by design and follows the current behavior of the set processor.